####-Pulse-Channels-IR-and-Compiler.md

```diff
@@ -0,0 +1,145 @@
+            Play(GaussianSquare(...), DriveChannel(4))
+```
+
+Note how one can't make sense of the code, without knowing the mapping of the control channels.
+Similarly, frame synchronization also has to be carried out manually and depends on the backend mapping.
+This dependency on the backend prohibits the ability to write a template code, which in turn complicates modules like Qiskit Experiments.
+
+Under the new proposal, the dependency on backend mapping will be removed, and Qiskit Pulse users (and particularly Qiskit Experiments users) will be able to write
+backend-agnostic template programs, which could be used across devices. Additionally, custom frames will make it easier to experiment with qudit control.
+
+The extra compilation needs introduced by the new channel model motivate the introduction the new Pulse IR and compiler. However, this was also
+a long standing goal on its own, further motivated by the desire to unify pulse compilation and allow for more flexibility in output formats.
+
+## User Benefit
+- Qiskit Pulse users will have a clearer and more efficient way to write their Pulse programs.
+- Qiskit Experiments users will be able to write more streamlined template experiments.
+- Qiskit Pulse users will have easier access to qudit control.
+- Maintainers and contributors to Qiskit Pulse will have a unified compilation code which will be simpler to adjust for future needs and changes.
+
+## Design Proposal
+One can schematically split the user interface of Qiskit Pulse into three layers -
+
+- Pulse level (`SymbolicPulse`,`WaveForm`)
+- Instruction level (`Play`,`Delay`...)
+- Program level (`Schedule`,`ScheduleBlock`)
+This proposal focuses on the instruction level, and the necessary compilation needs that will arise.
+The main goal is to allow for clear and simple backend agnostic (but not architecture agnostic) pulse templates.
+To do this, one must move away from the `Channel`s setup, which can't be used without specific backend mapping.
+
+The new Pulse IR and compiler will support the transition, and provide a unified compilation tool for every need and desired output format.
+
+No API breaking changes are needed. New and legacy options can coexist.
+
+## Detailed Design
+### Channel rework
+To replace the legacy Channels we propose to specify instructions in terms of `LogicalElement`s and `Frame`s:
+
+- `LogicalElement` - every type of element in the HW which the user can control ("play pulses on"). For example, a qubit is a `LogicalElement`, and not the specific port used to drive it. The most notable example of a logical element is of course the `Qubit`, but in the future one can imagine dedicated couplers, flux controls and so on.
+- `Frame` - a combination of frequency and phase. Subclasses used to identify the frame with a backend default. Notable examples are `QubitFrame` and `MeasurementFrame` associated with the default driving and measurement frequencies (respectively) of a qubit.
```

Comment:
I was thinking that all indices should be regarded as logical, and the user will then provide the compiler with a mapping from logical to physical qubits. If no mapping is provided, the logical qubits will be trivially mapped to the physical qubits.

Reply:
Fair enough. What about Channel indices? This is assumed to be physical?

Reply:
Yes. Because they are linked to the backend mapping, I think you have to treat them as physical.

Reply:
Since we canonicalize Channels and convert everything into mixed frame representation (you plan to expose `MixedFrame`), we may have the mixture of physical and virtual frames. If we don't have explicit distinction, we should have layout stage before the mixed frame conversion pass to ensure mixed frame are all physical in IR.

Reply:
That's a good point. The ordering should be considered carefully.